simplemma/language_detector.py

```diff
@@ -1,128 +1,139 @@
+    ) -> None:
+        self.lang = lang
+        self.greedy = greedy
+        self.dictionary_factory = dictionary_factory
+        self.token_sampler = token_sampler
+
+    def proportion_in_each_language(
+        self,
+        text: str,
+    ) -> Dict[str, float]:
+        """Determine which proportion of the text is in each of the target language(s)."""
+        tokens = self.token_sampler.sample_text(text)
+
+        total_tokens = len(tokens)
+        if total_tokens == 0:
+            return {"unk": 1}
+
+        dictionaries = self.dictionary_factory.get_dictionaries(self.lang)
+        known_tokens_count = dict.fromkeys(dictionaries, 0)
+        unknown_tokens_count = 0
         for token in tokens:
-            candidate = _return_lemma(
-                token, lang_dictionary, greedy=greedy, lang=lang_code
-            )
-            if candidate is not None:
-                in_target += 1
-        # compute results
-        found_ratio = in_target / total_tokens
-        myresults[lang_code] = found_ratio
-        unknown = 1 - found_ratio or 0.0
-        if myresults.get("unk") is None or unknown < myresults["unk"]:
-            myresults["unk"] = unknown
-    results = sorted(myresults.items(), key=itemgetter(1), reverse=True)
-    # post-processing
-    if len(results) > 1:
-        # switch unknown to the end
-        for i, item in enumerate(results):
-            if item[0] == "unk":
-                pair = results.pop(i)
-                results.append(pair)
-        # in case of ex-aequo use other token sampling to discriminate
-        if not greedy and results[0][1] == results[1][1]:
-            results = lang_detector(
-                text, lang=lang, greedy=True, token_sampler=backup_sampler
-            )
-    return results
+            token_found = False
+            for lang_code, lang_dictionary in dictionaries.items():
+                candidate = _return_lemma(
+                    token, lang_dictionary, self.greedy, lang_code
+                )
+                if candidate is not None:
+                    known_tokens_count[lang_code] += 1
+                    token_found = True
+            if not token_found:
+                unknown_tokens_count += 1
+
+        results: Dict[str, float] = dict(
+            (lang_code, token_count / total_tokens)
+            for (lang_code, token_count) in known_tokens_count.items()
+        )
+        results["unk"] = unknown_tokens_count / total_tokens
+        return results
+
+    def proportion_in_target_languages(
+        self,
+        text: str,
+    ) -> float:
+        return sum(
+            percentage
+            for (
+                lang_code,
+                percentage,
+            ) in self.proportion_in_each_language(text).items()
+            if lang_code != "unk"
+        )
+
+    def main_language(
+        self,
+        text: str,
+        token_samplers: List[TokenSampler] = [
+            MostCommonTokenSampler(),
```

Comment:
This was better before.

The class has a token sampler set at the constructor.
And this specific method can provide additional samplers for discrimination.

With this change the, main tokenizer is not used here which is odd.

Reply:
But you were resetting the class sampler which seems cumbersome, maybe there is a better way to do this?

Reply:
```python
    def main_language(
        self,
        text: str,
        additional_token_samplers: List[TokenSampler] = [
            RelaxedMostCommonTokenSampler()
        ],
    ) -> str:
        original_token_sampler = self.token_sampler

        for token_sampler in [self.token_sampler] + additional_token_samplers:
            self.token_sampler = token_sampler
            list_results = _as_list(self.proportion_in_each_language(text))
            if len(list_results) > 1 and list_results[0][1] != list_results[1][1]:
                return list_results[0][0]

        self.token_sampler = original_token_sampler

        return "unk"
```

The logic here is that there is `self.token_sampler` which is used by all methods.
Additionally, for the `main_language` method you can provide a list of `additional_token_samplers` to discriminate in case the class `self.token_sampler` produces an ex-aequo.

The reason to modify temporarily the `self.token_sampler` is that it's used by the `proportion_in_each_language` method.

I can see 2 alternatives:
* pass the `token_sampler` to each method as it was before. This is repetitive and kind of defeat the purpose of having the class.
* get rid of the class and leave just the functions. This makes the functions repetitive since you need to pass the same details to every function

Reply:
Fair enough, but the original token sampler was not restored in case of early return (`return list_results[0][0]`).

I made another suggestion below, what do you think?